### PR TITLE
Implement spender approvals and transfers

### PR DESCRIPTION
**Pull request by tgntr**
Implements spender approval primitives for the asset pallet and adds a `transfer_funds` extrinsic to the settlement pallet for spending approved allowances and same-identity direct transfers.

### new features

- Added spender allowance system — owners can approve other accounts to transfer fungible assets on their behalf
- Added `transfer_funds` to the settlement pallet — supports same-identity direct transfers and spender-approved transfers without going through the settlement engine
- Same-identity transfers for `transfer_asset`
- Added `transfer_nft` to the NFT pallet — transfer NFTs between accounts/portfolios with same routing logic as `transfer_funds`

### modified external API

- `transfer_asset` now uses `ensure_origin_call_permissions` and routes same-DID transfers directly instead of always creating a settlement instruction

### new external API

- Added `approve` extrinsic to the asset pallet — set, replace, or revoke a spender allowance for a given asset
- Added `transfer_funds` extrinsic to the settlement pallet — transfer assets with optional spender approval, defaults source to caller's account when `from` is `None`
- Added `allowance(owner, spender, asset_id)` Runtime API — query remaining spender allowance without a transaction
- Added transfer_nft extrinsic to the NFT pallet — transfer NFTs with same-DID direct / cross-DID settlement routing

### new events

- Added `Approval { owner, spender, asset_id, amount }` event — emitted on `approve` calls, not on spend

### other

- Added `Allowances` storage map to the asset pallet — maps `(owner, spender, asset_id)` to approved balance
- Added `InsufficientAllowance` error to the asset pallet
- Added `SenderSameAsReceiver` and `AllowancesNotSupportedForNFTs` errors to the settlement pallet
- Added `spend_allowance` public helper to the asset pallet — used by the settlement pallet to check and decrement allowances
- Infinite allowance (`Balance::MAX`) is never decremented on spend; allowance depleted to zero removes the storage entry
- Added tests for `approve`, `allowance` Runtime API, and `transfer_funds`  covering spender mode, same-identity transfers, cross-identity settlement, portfolio custody, NFT transfers, atomicity, and edge cases
- Added `PortfolioFnTrait` / `PortfolioFnConfig`
- Added `transfer_funds` / `transfer_funds_weight` to `SettlementFnTrait` — exposes routing logic for cross-pallet callers
- `simplified_nft_transfer` now accepts `Option<InstructionId>` (was required)
- Removed dead code: `transfer_asset_base_weight` benchmark, `transfer_asset_and_try_execute`, `transfer_nft_and_try_execute`

### CODE REVIEW

File: pallets/asset/src/lib.rs [hunk 1]
```diff
@@ -1718,6 +1733,51 @@ pub mod pallet {
...
             Self::base_reject_asset_transfer(origin, transfer_id)
         }
+
+        /// Set the allowance for `spender` to transfer up to `amount` of `asset_id` from
+        /// the caller's account.
+        ///
+        /// Replaces any existing allowance for this (owner, spender, asset_id) combination.
+        /// Setting `amount` to 0 revokes the allowance (removes the storage entry).
+        /// Setting `amount` to `Balance::MAX` grants an unlimited allowance that is never
+        /// decremented on spend.
```
Comment (Neopallium) [inline pallets/asset/src/lib.rs:1760]: Is this the standard for ERC20?

File: pallets/asset/src/lib.rs [hunk 2]
```diff
@@ -1718,6 +1733,51 @@ pub mod pallet {
...
+        /// * `MissingIdentity` — Caller's key is not linked to a DID.
+        #[pallet::call_index(37)]
+        #[pallet::weight(<T as Config>::WeightInfo::approve())]
+        pub fn approve(
+            origin: OriginFor<T>,
+            asset_id: AssetId,
+            spender: T::AccountId,
+            amount: Balance,
+        ) -> DispatchResult {
+            let caller_data = IdentityPallet::<T>::ensure_origin_call_permissions(origin)?;
```
Comment (Neopallium) [inline pallets/asset/src/lib.rs:1779]: I don't think we should be doing call permissions here.  When account keys hold the assets only that key has the authority over those assets.

File: pallets/settlement/src/lib.rs [hunk 3]
```diff
@@ -1530,10 +1536,123 @@ pub mod pallet {
...
+        // TODO: Implement portfolio custody check for portfolio sources.
+
+        // Resolve DIDs and determine transfer path.
+        let from_did = pallet_identity::Pallet::<T>::asset_holder_did(&resolved_from)?;
+        let to_did = pallet_identity::Pallet::<T>::asset_holder_did(&to)?;
+
+        let mut weight_meter = WeightMeter::max_limit(<T as Config>::WeightInfo::transfer_funds());
+
+        // Spender or same-identity: transfer directly. Otherwise: use settlement engine.
+        if is_spender || from_did == to_did {
```
Comment (Neopallium) [inline pallets/settlement/src/lib.rs]: The `base_transfer` can only be used when `from_did == to_did`.  The selection of transfer vs. settlement is only decided based on the from/to DIDs.
```suggestion
        if from_did == to_did {
```

**Review by Neopallium — CHANGES_REQUESTED** [review]

File: pallets/asset/src/lib.rs (hunk 2 above)
Comment (Neopallium) [reply pallets/asset/src/lib.rs:1779]: Based on our discussion we will keep using call permissions here.  Which means we should change `Asset.transfer_asset` to also use call permissions.

File: pallets/asset/src/lib.rs (hunk 1 above)
Comment (Neopallium) [reply pallets/asset/src/lib.rs:1760]: I would prefer that we don't have an "unlimited allowance" special case.  If they approve `MAX`, that is basically unlimited for most assets.

Comment (tgntr) [reply pallets/asset/src/lib.rs:1760]: yes https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/ERC20.sol#L294-L304

File: pallets/asset/src/lib.rs [hunk 4]
```diff
@@ -593,6 +601,14 @@ pub mod pallet {
         ValueQuery,
     >;
 
+    /// Maps (owner, spender, asset_id) to the approved allowance amount.
+    ///
+    /// A non-existent entry returns 0 (via `ValueQuery`), matching ERC-20 behavior.
+    /// When an allowance is revoked (set to 0), the entry is removed to bound storage growth.
+    #[pallet::storage]
+    pub type Allowances<T: Config> =
+        StorageMap<_, Blake2_128Concat, (T::AccountId, T::AccountId, AssetId), Balance, ValueQuery>;
```
Comment (Neopallium) [inline pallets/asset/src/lib.rs]: This should be a double map, so that the owner can easily query all current allowances.

Comment (Neopallium) [reply pallets/asset/src/lib.rs]: Could even use a NMap.

File: pallets/settlement/src/benchmarking.rs [hunk 5]
```diff
@@ -933,4 +934,47 @@ benchmarks! {
...
+            ONE_UNIT * 10,
+        )
+        .unwrap();
+
+        // Mandatory receiver affirmation + caller is receiver = two affirmations + execution.
+        Pallet::<T>::set_mandatory_receiver_affirmation(
+            bob.origin().into(),
+            AffirmationRequirement::Required,
+        )
+        .unwrap();
```
Comment (Neopallium) [inline pallets/settlement/src/benchmarking.rs]: The settlement can't execute.  This benchmark will not generate a weight function with the worse-case.

File: pallets/settlement/src/benchmarking.rs [hunk 6]
```diff
@@ -933,4 +934,47 @@ benchmarks! {
...
+
+        let from = Some(AssetHolder::try_from(alice.account().encode()).unwrap());
+        let to = AssetHolder::try_from(bob.account().encode()).unwrap();
```
Comment (Neopallium) [inline pallets/settlement/src/benchmarking.rs:1028]: Using accounts instead of portfolios will be cheaper (no custodianship checks).

File: pallets/nft/src/lib.rs [hunk 7]
```diff
@@ -740,6 +764,33 @@ impl<T: Config> Pallet<T> {
...
+        let _caller_data = IdentityPallet::<T>::ensure_origin_call_permissions(origin.clone())?;
+
+        let fund = Fund {
+            description: FundDescription::NonFungible(nfts),
+            memo,
+        };
+
+        let (_instruction_id, consumed) =
+            <T as pallet_asset::Config>::SettlementFn::transfer_funds(
+                origin,
```
Comment (Neopallium) [inline pallets/nft/src/lib.rs]: The call permission check here is a duplicate, since it is checked in `transfer_funds` we don't need it here.
The pallet/extrinsic names are set before the extrinsic is called, the permissions checks are not based on the `fn` name.

```suggestion
        let fund = Fund {
            description: FundDescription::NonFungible(nfts),
            memo,
        };

        let (_instruction_id, consumed) =
            <T as pallet_asset::Config>::SettlementFn::transfer_funds(
                origin,
```

File: pallets/asset/src/lib.rs [hunk 8]
```diff
@@ -2713,24 +2805,23 @@ impl<T: AssetConfig> Pallet<T> {
         #[cfg(feature = "runtime-benchmarks")] bench_base_weight: bool,
     ) -> DispatchResultWithPostInfo {
         let from = ensure_signed(origin.clone())?;
-        let mut weight_meter = <T as Config>::SettlementFn::transfer_and_try_execute_weight_meter(
-            <T as Config>::WeightInfo::transfer_asset_base_weight(),
-            true,
-        );
 
-        // Create the transfer instruction via the settlement engine and affirm it as the sender.
-        let instruction_id = T::SettlementFn::transfer_asset_and_try_execute(
```
Comment (Neopallium) [inline pallets/asset/src/lib.rs:2724]: Since this method emits an event, we should still create the weight meter here.  Also since this code path only uses account holder (no portfolios), it's worse-case weight should be lower.

File: pallets/settlement/src/lib.rs [hunk 9]
```diff
@@ -3804,14 +3972,27 @@ impl<T: Config> Pallet<T> {
...
+
         // Affirm the instruction on behalf of the sender.
-        Self::unsafe_affirm_instruction(
-            origin_data.primary_did,
-            instruction_id,
-            [from].into(),
-            origin_data.secondary_key.as_ref(),
-            None,
-        )?;
+        Self::unsafe_affirm_instruction(from_did, instruction_id, [from].into(), sender_sk, None)?;
```
Comment (Neopallium) [inline pallets/settlement/src/lib.rs:4072]: This is very confusing.  The `spend_allowance` check is outside of this function and it is not clear how that check matches up with the logic here.

Please move the spender/portfolio check into a function that can be called here for the "different DIDs, use settlement" case.

File: pallets/settlement/src/lib.rs [hunk 10]
```diff
@@ -3888,9 +4069,17 @@ impl<T: Config> SettlementFnTrait<T> for Pallet<T> {
         weight_meter: &mut WeightMeter,
         #[cfg(feature = "runtime-benchmarks")] bench_base_weight: bool,
     ) -> Result<Option<InstructionId>, DispatchErrorWithPostInfo> {
```
Comment (Neopallium) [inline pallets/settlement/src/lib.rs]: I think this is dead code now.

File: pallets/settlement/src/lib.rs [hunk 11]
```diff
@@ -1530,10 +1539,159 @@ pub mod pallet {
...
+                        nft_id: nfts.ids()[0],
+                    }
+                }
+            };
+            Self::base_transfer_and_try_execute(
+                origin,
+                Some(resolved_from),
+                Some(from_did),
+                to,
+                Some(to_did),
```
Comment (Neopallium) [inline pallets/settlement/src/lib.rs]: I think this is the only call to that function, so don't need optional parameters.

File: pallets/settlement/src/lib.rs [hunk 12]
```diff
@@ -1530,10 +1539,159 @@ pub mod pallet {
...
+                            return Err(Error::<T>::AllowancesNotSupportedForNFTs.into());
+                        }
+                    }
+                }
+            }
+            // Portfolio source: caller must have custody.
+            AssetHolder::Portfolio(ref portfolio_id) => {
+                T::PortfolioFn::ensure_portfolio_custody(portfolio_id, caller_did)?;
+            }
+        }
```
Comment (Neopallium) [inline pallets/settlement/src/lib.rs]: Please move this spender/custody check into a function, so it can be called in `base_transfer_and_try_execute` or inside the "Same-identity: transfer directly" block below.

Also I think we only need to do the spender check if `from.is_some()`.

**Review by Neopallium — CHANGES_REQUESTED** [review]

File: pallets/weights/src/pallet_settlement.rs [hunk 13]
```diff
@@ -1342,4 +1341,11 @@ impl pallet_settlement::WeightInfo for SubstrateWeight {
             .saturating_add(DbWeight::get().reads(1))
             .saturating_add(DbWeight::get().writes(1))
     }
+    // TODO: Replace with generated weights after benchmark run.
+    // Based on base_transfer (21r/5w) + origin check (1r) + allowance (1r/1w) + DID resolution (2r).
+    fn transfer_funds() -> Weight {
```
Comment (Neopallium) [inline pallets/weights/src/pallet_settlement.rs]: Please run new benchmarks locally (use a placeholder weight function during the first build/run), since the benchmarks will record storage read/writes.

File: pallets/settlement/src/benchmarking.rs [hunk 14]
```diff
@@ -933,4 +934,47 @@ benchmarks! {
     set_mandatory_receiver_affirmation {
         let alice = UserBuilder::<T>::default().generate_did().build("Alice");
     }: _(alice.origin, AffirmationRequirement::Required)
+
+    transfer_funds {
```
Comment (Neopallium) [inline pallets/settlement/src/benchmarking.rs]: I think we will need to split this benchmark/weight function into multiple functions.

Benchmarks for the main code paths (with receiver affirmation, no execution):

1. sender or receiver is a portfolio, same DID
2. sender or receiver is a portfolio, different DID
3. sender and receiver accounts, same DID
4. sender and receiver accounts, different DID

Also a benchmark to cover (

Comment (Neopallium) [reply pallets/settlement/src/benchmarking.rs]: My last comment got cut off.

Also a benchmark to cover the try execution logic.

File: pallets/weights/src/pallet_settlement.rs [hunk 15]
```diff
@@ -1342,4 +1341,28 @@ impl pallet_settlement::WeightInfo for SubstrateWeight {
             .saturating_add(DbWeight::get().reads(1))
             .saturating_add(DbWeight::get().writes(1))
     }
+    fn transfer_funds_portfolio_same_did() -> Weight {
+        // Minimum execution time: 141_000_000 picoseconds.
+        Weight::from_parts(142_000_000, 0)
+            .saturating_add(DbWeight::get().reads(15))
+            .saturating_add(DbWeight::get().writes(4))
+    }
```
Comment (Neopallium) [inline pallets/weights/src/pallet_settlement.rs:1377]: Please run the benchmarks locally and commit the new weight functions including function docs (with the description of what storage items are read/written).  I like to review the storage reads/writes against what I expect the code to do.

File: pallets/asset/src/lib.rs [hunk 16]
```diff
@@ -1638,7 +1665,7 @@ pub mod pallet {
         /// * `UnexpectedOFFChainAsset` - If the asset could not be found on-chain.
         /// * `MissingIdentity` - The caller doesn't have an identity.
         #[pallet::call_index(34)]
-        #[pallet::weight(<T as Config>::SettlementFn::transfer_and_try_execute_weight_meter(<T as Config>::WeightInfo::transfer_asset_base_weight(), true).limit())]
+        #[pallet::weight(<T as Config>::SettlementFn::transfer_funds_weight())]
```
Comment (Neopallium) [inline pallets/asset/src/lib.rs]: The weight here should be based the case where the sender/receiver can only be accounts.

File: pallets/asset/src/lib.rs [hunk 17]
```diff
@@ -2712,25 +2813,28 @@ impl<T: AssetConfig> Pallet<T> {
         memo: Option<Memo>,
         #[cfg(feature = "runtime-benchmarks")] bench_base_weight: bool,
     ) -> DispatchResultWithPostInfo {
-        let from = ensure_signed(origin.clone())?;
-        let mut weight_meter = <T as Config>::SettlementFn::transfer_and_try_execute_weight_meter(
-            <T as Config>::WeightInfo::transfer_asset_base_weight(),
-            true,
-        );
+        let from = frame_system::ensure_signed(origin.clone())?;
```
Comment (Neopallium) [inline pallets/asset/src/lib.rs:2823]: Since it was decided to do key call permission checks in the `approve` extrinsic, then we should do the same here.

File: pallets/settlement/src/benchmarking.rs [hunk 18]
```diff
@@ -933,4 +934,94 @@ benchmarks! {
...
+    verify {
+        assert_eq!(pallet_asset::Pallet::<T>::get_holders_balance(&to, &asset_id), ONE_UNIT);
+    }
+
+    transfer_funds_portfolio_diff_did {
+        let alice = UserBuilder::<T>::default().generate_did().build("Alice");
+        let bob = UserBuilder::<T>::default().generate_did().build("Bob");
+
+        let (sender_holding, receiver_holding, _, asset_id) =
+            setup_asset_transfer::<T>(&alice, &bob, None, None, true, true, 0, true, false);
```
Comment (Neopallium) [inline pallets/settlement/src/benchmarking.rs]: For difference DIDs (which will use a settlement), we need to include compliance and transfer restrictions for this weight function to be used for the worse-case (extrinsic estimated weight).
```suggestion
        let (sender_holding, receiver_holding, _, asset_id) =
            setup_asset_transfer::<T>(&alice, &bob, None, None, false, false, 0, true, false);
```

File: pallets/settlement/src/benchmarking.rs [hunk 19]
```diff
@@ -933,4 +934,94 @@ benchmarks! {
...
+        assert_eq!(pallet_asset::Pallet::<T>::get_holders_balance(&to, &asset_id), ONE_UNIT);
+    }
+
+    transfer_funds_account_diff_did {
+        let alice = UserBuilder::<T>::default().generate_did().build("Alice");
+        let bob = UserBuilder::<T>::default().generate_did().build("Bob");
+        let charlie = UserBuilder::<T>::default().generate_did().build("Charlie");
+
+        let (_sender_holding, _receiver_holding, _, asset_id) =
+            setup_asset_transfer::<T>(&alice, &charlie, None, None, true, true, 0, false, true);
```
Comment (Neopallium) [inline pallets/settlement/src/benchmarking.rs]: Same here.
```suggestion
        let (_sender_holding, _receiver_holding, _, asset_id) =
            setup_asset_transfer::<T>(&alice, &charlie, None, None, false, false, 0, false, true);
```

File: pallets/settlement/src/benchmarking.rs [hunk 20]
```diff
@@ -933,4 +934,94 @@ benchmarks! {
...
+    }
+
+    transfer_funds_account_same_did {
+        let alice = UserBuilder::<T>::default().generate_did().build("Alice");
+        let bob = UserBuilder::<T>::default().generate_did().build("Bob");
+
+        let (_sender_holding, _receiver_holding, _, asset_id) =
+            setup_asset_transfer::<T>(&alice, &bob, None, None, true, true, 0, false, true);
+
+        let from = AssetHolder::try_from(alice.account().encode()).unwrap();
```
Comment (Neopallium) [inline pallets/settlement/src/benchmarking.rs]: Need to include the spender approval cost in this benchmark.

File: pallets/settlement/src/lib.rs [hunk 21]
```diff
@@ -1530,10 +1555,173 @@ pub mod pallet {
...
+        ensure!(resolved_from != to, Error::<T>::SenderSameAsReceiver);
+
+        // Resolve DIDs and determine transfer path.
+        let from_did = pallet_identity::Pallet::<T>::asset_holder_did(&resolved_from)?;
+        let to_did = pallet_identity::Pallet::<T>::asset_holder_did(&to)?;
+
+        let same_did = from_did == to_did;
+        let is_portfolio = matches!(resolved_from, AssetHolder::Portfolio(_));
+        let _ =
+            weight_meter.check_accrue(Self::transfer_funds_actual_weight(is_portfolio, same_did));
```
Comment (Neopallium) [inline pallets/settlement/src/lib.rs]: Please don't ignore errors.

File: pallets/settlement/src/lib.rs [hunk 22]
```diff
@@ -3529,6 +3717,15 @@ impl<T: Config> Pallet<T> {
...
     }
 
+    fn transfer_funds_actual_weight(is_portfolio: bool, same_did: bool) -> Weight {
+        match (is_portfolio, same_did) {
+            (true, true) => <T as Config>::WeightInfo::transfer_funds_portfolio_same_did(),
+            (true, false) => <T as Config>::WeightInfo::transfer_funds_portfolio_diff_did(),
+            (false, true) => <T as Config>::WeightInfo::transfer_funds_account_same_did(),
+            (false, false) => <T as Config>::WeightInfo::transfer_funds_account_diff_did(),
+        }
+    }
```
Comment (Neopallium) [inline pallets/settlement/src/lib.rs]: Right now the benchmarks for these weights are not including compliance/transfer checks, so they are not covering the worse-case weight.

This function is used to "charge" the weight and shouldn't include the compliance/transfer checks overhead, since those will be charged to the meter (different DIDs) when the settlement executes.

With the weights in this current PR and asset with compliance/transfer rules would fail to be transferred between holders with different DIDs.

**Review by Neopallium — CHANGES_REQUESTED** [review]

File: pallets/asset/src/lib.rs (hunk 17 above)
Comment (tgntr) [reply pallets/asset/src/lib.rs:2823]: the `ensure_origin_call_permissions` does happen in `base_transfer_funds` which we call in this function

File: pallets/settlement/src/lib.rs [hunk 23]
```diff
@@ -1530,10 +1560,177 @@ pub mod pallet {
...
+
+    /// Authorize the transfer source.
+    ///
+    /// - Account source where caller != owner: checks and decrements spender allowance.
+    /// - Portfolio source: checks custody.
+    /// Authorize the transfer source.
+    ///
+    /// - Account source where caller != owner: checks and decrements spender allowance.
+    /// - Portfolio source: checks custody.
```
Comment (Neopallium) [inline pallets/settlement/src/lib.rs]: Duplicated docs.
```suggestion
    /// Authorize the transfer source.
    ///
    /// - Account source where caller != owner: checks and decrements spender allowance.
    /// - Portfolio source: checks custody.
```

File: pallets/nft/src/lib.rs [hunk 24]
```diff
@@ -297,6 +298,29 @@ pub mod pallet {
...
+        /// * `origin` — Signed origin. Caller must have a registered DID.
+        /// * `nfts` — The NFTs to transfer.
+        /// * `from` — Source. `None` defaults to caller's account.
+        /// * `to` — Destination account or portfolio.
+        /// * `memo` — Optional memo attached to the transfer.
+        #[pallet::call_index(4)]
+        #[pallet::weight(<T as pallet_asset::Config>::SettlementFn::transfer_funds_weight())]
+        pub fn transfer_nft(
+            origin: OriginFor<T>,
+            nfts: NFTs,
```
Comment (Neopallium) [inline pallets/nft/src/lib.rs]: This weight function needs to take into account the number of NFTs being transferred here (since `NFTs` can reference multiple NFTs in the same collection).

File: pallets/nft/src/lib.rs [hunk 25]
```diff
@@ -297,6 +298,29 @@ pub mod pallet {
...
+            from: Option<AssetHolder>,
+            to: AssetHolder,
+            memo: Option<Memo>,
+        ) -> DispatchResultWithPostInfo {
```
Comment (Neopallium) [inline pallets/nft/src/lib.rs]: To keep this one simple lets restrict it to only `Account` like we do with the `Asset.transfer` extrinsic.  Anyone that wants to use different kinds of from/to can use the `Settlement.transfer_funds` extrinsic.

This will also help keep the worse-case cost down (which is also the minimum POLYX the caller needs hold to submit the tx).  We have had users complain about high worse-case estimated tx fee costs before.

File: pallets/settlement/src/lib.rs [hunk 26]
```diff
@@ -1530,10 +1560,177 @@ pub mod pallet {
...
+        #[pallet::call_index(26)]
+        #[pallet::weight(<T as Config>::WeightInfo::transfer_funds())]
+        pub fn transfer_funds(
+            origin: OriginFor<T>,
+            from: Option<AssetHolder>,
+            to: AssetHolder,
+            fund: Fund,
+        ) -> DispatchResultWithPostInfo {
+            let mut weight_meter =
+                WeightMeter::max_limit(<T as Config>::WeightInfo::transfer_funds());
```
Comment (Neopallium) [inline pallets/settlement/src/lib.rs]: The weight function here will need to detect (from the `fund` value) if it is an asset or NFTs, since the weight function needs to be based on the number of NFTs.

File: pallets/settlement/src/lib.rs [hunk 27]
```diff
@@ -1530,10 +1560,177 @@ pub mod pallet {
...
+        let to_did = pallet_identity::Pallet::<T>::asset_holder_did(&to)?;
+
+        let same_did = from_did == to_did;
+        let is_portfolio = matches!(resolved_from, AssetHolder::Portfolio(_));
+
+        // Charge the benchmark-measured weight for this path.
+        Self::check_accrue(
+            weight_meter,
+            Self::transfer_funds_actual_weight(is_portfolio, same_did),
+        )?;
```
Comment (Neopallium) [inline pallets/settlement/src/lib.rs:1670]: I was wrong about needing to include the compliance/transfer rule cost in these benchmarks.  But for now this should cover it.  We only have to handle the NFTs case in the benchmarks and weight functions.

**Review by HenriqueNogara — APPROVED** [review]

**Review by Neopallium — CHANGES_REQUESTED** [review]
Overview of changes for this PR:

1. Update the benchmarks/weights for NFTs.
2. `Nft.transfer_nft` use `AccountId` instead of `AssetHolder` for from/to.

For follow-up PR:

1. Use the `fund`, `from`, `to` parameters to select a more optimal worse-case weight (to lower the estimated tx fee).
2. Don't include compliance/transfer rule and execution cost in base weights, so that the meter can be used for calculating a better actual fee.

File: pallets/settlement/src/lib.rs [hunk 28]
```diff
@@ -4102,7 +4143,8 @@ impl<T: Config> Pallet<T> {
         )?;
 
         // Try to execute the instruction.
-        Self::base_try_execute_instruction(origin, instruction_id, is_fungible, weight_meter)?;
+        let asset_count = Self::fungible_asset_count(is_fungible);
+        Self::base_try_execute_instruction(origin, instruction_id, &asset_count, weight_meter)?;
```
Comment (Neopallium) [inline pallets/settlement/src/lib.rs:4147]: Right now the `is_fungible` flag is only set to `true`, since this method is only called for the fungible case.  If we decide to add a `Nft.receiver_affirm_nft_transfer` extrinsic, then we will need to have the `AssetCount` include the correct number of NFTs.

For this PR, this is fine as is.

**Review by Neopallium — APPROVED** [review]

**Review by HenriqueNogara — APPROVED** [review]
